File: H1_SDK1.3.6/camera_sdk_cpp/examples/03_multi_cam_viewer.cpp
/**
 * @file 03_multi_cam_viewer.cpp
 * @brief 【示例 03】自动发现多相机流
 * * [功能介绍]
 * 自动从服务端获取所有可用的相机 ID，无需手动硬编码相机名称。
 * * [使用说明]
 * 1. 修改 GRPC_TARGET 为服务端真实的 IP 地址和端口（例如 "192.168.1.100:50051"）。
 * 2. 程序会自动调用 get_state() 接口拉取配置列表。
 * * [注意]
 * 如果该接口返回列表为空，说明服务端未成功挂载任何相机设备。
 */

#include <iostream>
#include <vector>
#include <string>
#include <opencv2/opencv.hpp>
#include "camera_client.h"

int main() {
  // ---------------------------------------------------------
  // 【用户配置区】请根据实际环境修改以下参数
  // ---------------------------------------------------------
  const std::string GRPC_TARGET = "localhost:50051";
  // ---------------------------------------------------------

  CameraClient client(GRPC_TARGET);
  client.start();

  std::vector<std::string> camera_names;
  try {
    // 自动发现：获取所有可用相机
    auto state = client.get_state({}, 5.0);
    for (int i = 0; i < state.camera_configs_size(); ++i) {
      camera_names.push_back(state.camera_configs(i).camera_name());
    }
  } catch (const std::exception& e) {
    std::cerr << "[ERROR] 获取状态失败，请检查 IP [" << GRPC_TARGET << "]: " << e.what() << std::endl;
    client.stop();
    return -1;
  }

  std::cout << "[INFO] 正在开启多窗口预览，按 'q' 退出..." << std::endl;
  while (true) {
    for (const auto& name : camera_names) {
      auto frame = client.getLatestFrame(name);
      if (frame) {
        cv::imshow("Stream: " + name, frame->first);
      }
    }

    if (cv::waitKey(1) == 'q') break;
  }

  cv::destroyAllWindows();
  client.stop();
  return 0;
}

File: H1_SDK_1.3.7/camera_sdk_python/example/03_multi_cam_viewer.py
"""
@file 03_multi_cam_viewer.py
@brief 【示例 03】多相机自动发现与流显示
* [功能介绍]
1. 调用 get_state 接口动态查询当前服务端所有在线相机。
2. 自动为每个相机开启对应的预览窗口。
* [使用说明]
1. 仅需配置 GRPC_TARGET。
2. 适合用于验证系统内所有相机是否工作正常。
"""

import os
import sys
import cv2

sys.path.append(os.path.join(os.path.dirname(__file__), '..'))
from camera_client import CameraClient

def main():
    GRPC_TARGET = "localhost:50051"
    client = CameraClient(grpc_target=GRPC_TARGET)
    client.start()

    # 1. 动态发现相机
    try:
        state = client.get_state()
        # 提取所有相机的名称
        camera_names = [config.camera_name for config in state.camera_configs]
        print(f"[+] 发现相机列表: {camera_names}")
    except Exception as e:
        print(f"[-] 获取状态失败: {e}")
        client.stop()
        return

    try:
        while True:
            for name in camera_names:
                # 获取每一路的彩色帧
                frame_data = client.get_latest_frame(name)
                if frame_data:
                    img, _ = frame_data
                    cv2.imshow(f"Camera: {name}", img)
            
            if cv2.waitKey(1) & 0xFF == ord('q'):
                break
    finally:
        client.stop()
        cv2.destroyAllWindows()

if __name__ == "__main__":
    main()